### ISSUE TITLE
Python switcheroo causes Munki 5.1 self-upgrade issues

### ISSUE THREAD

**Issue by gregneagle** [munki/munki#1014]
From @MagerValp on MacAdmins Slack:

> installing 5.1b1 with munki itself pulls /usr/local/munki/python out from under itself
> `autopkg run -k INCLUDE_PRERELEASES=true munkitools5.munki` and then deploy that to a client running 5.0.1
> Seems like there needs to be additional logic if you want it to work when broken up into component packages

This is related to this change: https://github.com/munki/munki/pull/997 and this one: https://github.com/munki/munki/commit/19b5555d7e6baece51c0964679e195a5bf59533f

> Which causes /usr/local/munki/python to be deleted
> And since the new python 3.8.5 package is the first thing munki installs
> It effectively kills itself before it can install the updated managedsoftwareupdate
> Leading to a broken munki install

**Comment by gregneagle** [comment]
I'm thinking the best solution might be: updated Python pkg leaves the 3.7 version directory and python symlink in place, but `managedsoftwareupdate` on the next execution does the cleanup. The exact mechanism for _that_ could be one of several things: a cleanup script dropped by the updated Python pkg which `managedsoftwareupdate` runs; some data stored in /Library/Preferences/ManagedInstalls.plist that `managedsoftwareupdate` reads and acts upon, or something else...

**Comment by MagerValp** [comment]
I did a little experimentation and installed 4.0, upgraded python to 3.8.5, and created a `/usr/local/munki/python` symlink. 4.0 seems to run fine on 3.8.5, though it's getting a few warnings about buffering. It runs well enough to update itself with the 5.1b1 component packages, after which it's possible to delete the symlink.

```✔ # managedsoftwareupdate --version
4.0.0.3881
✔ # /usr/local/munki/python -V       
Python 3.8.5
✔ # managedsoftwareupdate --auto     
/usr/local/munki/Python.framework/Versions/3.8/lib/python3.8/subprocess.py:838: RuntimeWarning: line buffering (buffering=1) isn't supported in binary mode, the default buffer size will be used
  self.stdin = io.open(p2cwrite, 'wb', bufsize)
/usr/local/munki/Python.framework/Versions/3.8/lib/python3.8/subprocess.py:844: RuntimeWarning: line buffering (buffering=1) isn't supported in binary mode, the default buffer size will be used
  self.stdout = io.open(c2pread, 'rb', bufsize)
/usr/local/munki/Python.framework/Versions/3.8/lib/python3.8/subprocess.py:849: RuntimeWarning: line buffering (buffering=1) isn't supported in binary mode, the default buffer size will be used
  self.stderr = io.open(errread, 'rb', bufsize)
/usr/local/munki/Python.framework/Versions/3.8/Resources/Python.app/Contents/MacOS/../../../../../../Versions/3.8/lib/python3.8/codecs.py:905: RuntimeWarning: line buffering (buffering=1) isn't supported in binary mode, the default buffer size will be used
  file = builtins.open(filename, mode, buffering)
/usr/local/munki/munkilib/scriptutils.py:39: RuntimeWarning: line buffering (buffering=1) isn't supported in binary mode, the default buffer size will be used
  fileobject = open(path, mode='wb', buffering=1)
✔ # installer -target / -pkg ~/Downloads/munkitools_core-5.1.0.4072.pkg 
installer: Package name is munkitools_core-5.1.0.4072
installer: Upgrading at base path /
installer: The upgrade was successful.
✔ # installer -target / -pkg ~/Downloads/munkitools_app-5.2.0.4071.pkg 
installer: Package name is munkitools_app-5.2.0.4071
installer: Upgrading at base path /
installer: The upgrade was successful.
✔ # installer -target / -pkg ~/Downloads/munkitools_admin-5.1.0.4072.pkg 
installer: Package name is munkitools_admin-5.1.0.4072
installer: Upgrading at base path /
installer: The upgrade was successful.
✔ # installer -target / -pkg ~/Downloads/munkitools_app_usage-5.1.0.4072.pkg 
installer: Package name is munkitools_app_usage-5.1.0.4072
installer: Upgrading at base path /
installer: The upgrade was successful.
✔ # managedsoftwareupdate --version
5.1.0.4072
✔ # managedsoftwareupdate --auto   
✔ # rm -f python
✔ # 
```

So in conclusion, I don't think it's critical to keep the 3.7 framework, only the symlink, if that makes switching easier.

**Comment by MagerValp** [comment]
`munkitools_python-3.8.5.plist`:
```
        <key>postinstall_script</key>
        <string>
cd /usr/local/munki
ln -s Python.framework/Versions/Current/bin/python3 python
</string>
```

`munkitools_core-5.1.0.4072.plist`:
```
        <key>postinstall_script</key>
        <string>
rm -f /usr/local/munki/python
</string>
```

This was _not_ successful. It hangs after installing the package, and I had to ctrl-c:

``` # managedsoftwareupdate -v --installonly
Managed Software Update Tool
Copyright 2010-2020 The Munki Project
https://github.com/munki/munki

Starting...
    Preventing idle sleep
Installing Managed Software Center embedded python (1 of 5)...
    Förbereder installation…
    Förbereder skivan…
    Förbereder munkitools_python-3.8.5…
    Väntar på att andra installationer ska avslutas…
    Anpassar installationen…
    3.175 percent complete...
    Kontrollerar paket…
    9.617703 percent complete...
    Installerar filer…
    14.936891 percent complete...
    Installerar filer…
    20.256079 percent complete...
    Installerar filer…
    25.575267 percent complete...
    Installerar filer…
    28.234861 percent complete...
    Installerar filer…
    53.512754 percent complete...
    Registrerar paket i system…
    83.402671 percent complete...
    Kör paketskript…
    84.37025 percent complete...
    Kör paketskript…
    85.537155 percent complete...
    Skriver paketkvitton…
    89.771218 percent complete...
    Optimerar systemet för installerad programvara…
    Kontrollerar paket…
    97.75 percent complete...
    Avslutar installationen…
    100.0 percent complete...
    Programvaran har installerats.
^CERROR: Unexpected error in munkilib.installer:
Traceback (most recent call last):
  File "/usr/local/munki/managedsoftwareupdate", line 1197, in <module>
    main()
  File "/usr/local/munki/managedsoftwareupdate", line 1016, in main
    restart_action = doInstallTasks(appleupdatesavailable)
  File "/usr/local/munki/managedsoftwareupdate", line 244, in doInstallTasks
    only_unattended=only_unattended)
  File "/usr/local/munki/munkilib/installer/core.py", line 704, in run
    only_unattended=only_unattended))
  File "/usr/local/munki/munkilib/installer/core.py", line 321, in install_with_info
    item, itempath)
  File "/usr/local/munki/munkilib/installer/core.py", line 197, in handle_apple_package_install
    (retcode, needtorestart) = pkg.install(itempath, item)
  File "/usr/local/munki/munkilib/installer/pkg.py", line 259, in install
    retcode = _run_installer(cmd, env_vars, packagename)
  File "/usr/local/munki/munkilib/installer/pkg.py", line 178, in _run_installer
    time.sleep(1)
KeyboardInterrupt
    Allowing idle sleep
```

Manually creating the symlink and re-running allows it to finish successfully, and the symlink is cleaned up by the core package without interruption.

**Comment by MagerValp** [comment]
The last idea I have down this path would be to add the symlink creation to the python package `postinstall` (as opposed to the pkgsinfo) and see if that is enough to keep it from tripping. I've run out of time for today though.

**Comment by gregneagle** [comment]
"The last idea I have down this path would be to add the symlink creation to the python package postinstall" It's currently the python pkg postinstall that _removes_ the /usr/local/munki/python symlink...

**Comment by natewalck** [comment]
Instead of doing it all in one run would it be worth having this cleanup done in a simple pkginfo that does nothing but clean up this legacy python symlink? We could then use a conditional to scope the cleanup pkginfo to Version 5.1.x or later to ensure it doesn't yank it before we are on the new stuff.

**Comment by gregneagle** [comment]
What about scenarios where _Munki_ isn't doing the install? The package should really do the right thing...

**Comment by natewalck** [comment]
That is true. How about if in the preflight we check for the new Munki-python and clean it up only then. 

This means they wouldn't get the cleanup until the next Munki update after the python swap, but it would keep it in the package and be more or less safe.

If someone *really* wants it cleaned up immediately, they could do that outside of Munki as well.

**Comment by natewalck** [comment]
Took a crack at this. Its not perfect, but it does handle this situation in a safer fashion (but not as immediate).

**Comment by gregneagle** [comment]
An alternative would be yet another component pkg that did the cleanup. It would be selected (and installed) by default. For Munki, the recommendation would be to _not_ install that component pkg and instead handle the cleanup seperately.
Any of these solutions seem terribly complex, though.
I think longer-term the right approach might be to hand off Munki upgrades to a process other than `managedsoftwareupdate`

**Comment by natewalck** [comment]
I thought about the component package as well, but realized it didn't get us a one-pass update and cleanup. *Not* installing it is fine, I guess it would make it easier to do the cleanup with munki itself or some other tool after the initial switcheroo.

I was thinking of a secondary process to do the cleanup, but I like the idea of a helper process that does the Munki update itself outside of `managedsoftwareupdate`. The only thing that sucks there is it probably shouldn't be in Python, otherwise we may hit the same update issues. This feels like a one time thing though (changing `python` to `munki-python`), so maybe all of this is premature optimization.

**Comment by natewalck** [comment]
Any thoughts on this? Would you prefer if I break this mechanism out into a different sub-package or do we want to go an entirely different direction and have something like an `UpdateSupervisor` launchd that handles updates for Munki itself.

**Comment by gregneagle** [comment]
I looked at this today. I'm still not entirely happy with any of the proposed solutions. Even a dedicated process to self-update Munki is problematic -- it's probably a good idea longer term, but won't really fix this particular problem because none of the affected machines would yet have this proposed process in place!

**Comment by gregneagle** [comment]
I'm seriously considering backing out the change from `/usr/local/munki/python` to `/usr/local/munki/munki-python` for now until we have a better handle on managing this change.

**Comment by natewalck** [comment]
I was thinking the exact same thing with reverting this. It is causing a fair amount of headaches and I think if someone really wants to avoid this issue, they could remove `/usr/local/munki/python` from the global paths, correct?

**Comment by gregneagle** [comment]
I'd think they'd have to remove /usr/local/munki from the PATH.

**Comment by natewalck** [comment]
Right, I didn't remove the symlink :D You'd use the directory it lives in.

**Comment by gregneagle** [comment]
I've reverted the `/usr/local/munki/python` to `/usr/local/munki/munki-python` change for now.

**Comment by gregneagle** [comment]
The current pkg built from the Munki5dev branch leaves the /usr/local/munki/python symlink, but the postflight in the python subpackage still removes the Python.framework/Versions/3.7 directory.

In some brief testing, this still causes Munki to hang when upgrading itself immediately after the install of the python subpackage.

So a variation of this problem still exists and we cannot move forward with a 5.1 release without a resolution. Munki clients must be able to successfully upgrade themselves from an older version of Munki to version 5.1 (or later).

**Comment by gregneagle** [comment]
@natewalck ^

**Comment by natewalck** [comment]
Maybe we should go back to the earlier option of doing cleanup in the preflight only. You'd have 2 versions of python on disk until the next Munki package install or the admin can clean it up themselves. It seems like the **safest** way to proceed, but it isn't necessarily the most space efficient.

Or we can get Rube Golbergy on this and have the postlfight touch a flag file on disk with a launchd watch path, sleeps for N seconds and then nuke the old Python. This idea feels gross to me.

**Comment by gregneagle** [comment]
OK, we're now at this point:

Munki5dev code produces a package that installs the new Python.framework and munki-python symlink, but does no cleanup. Early testing indicates Munki can self-upgrade using this package without crashing. So what remains is to do cleanup of the old Python 3.7 version and the /usr/local/munki/python symlink _after_ the current `managedsoftwareupdate` process exits.  

This brings us back to https://github.com/munki/munki/issues/1014#issuecomment-680596723

**Comment by natewalck** [comment]
Graham has a solid generalized solution for this in #1025

**Comment by gregneagle** [comment]
I believe this is addressed in Munki 5.1 Beta 2.
